README.md

```diff
@@ -279,12 +279,13 @@ Contributions of valid and invalid OpenDrive sample files are also welcome. New
 1. Create a new Python module for each checker.
 2. Specify the following global variables for the Python module
 
-| Variable | Meaning |
-| --- | --- |
-| `CHECKER_ID` | The ID of the checker |
-| `CHECKER_DESCRIPTION` | The description of the checker |
-| `CHECKER_PRECONDITIONS` | A set of other checkers in which if any of them raise an issue, the current checker will be skipped |
-| `RULE_UID` | The rule UID of the rule that the checker will check |
+| Variable | Presence | Meaning |
+| --- | --- | --- |
+| `CHECKER_ID` | Required | The ID of the checker |
+| `CHECKER_DESCRIPTION` | Required | The description of the checker |
+| `CHECKER_PRECONDITIONS` | Required | A set of other checkers in which if any of them raise an issue, the current checker will be skipped |
+| `RULE_UID` | Required | The rule UID of the rule that the checker will check |
+| `LAST_SUPPORTED_VERSION` | Optional | The last supported version of the standard. If the input file has a version higher than the last supported version, the checker will be skipped. Do not define this variable if the checker does not have a last supported version. |
```

Comment:
Don't like the term "supported version" in this context, because we don't "support" rules. It's more that the there is a highest version which requires this rule.

But don't have a good term at the moment.... 

Any ideas by anybody?

Reply:
I would also go with one of the proposal from @andreaskern74, i.e. "LAST_APPLICABLE_VERSION"
If there is no one against it.
So we can close this PR and create the two remaining "release candidates", because I would like to provide the link to this release candidates as soon as possible to the ASAM TSC

Reply:
We have to take into consideration that:

 * the UID already contains an hint for the version at which a rule should be applied.
 * all the rules are actually defined _after_ the emanation of the standard. _(I can provideas an example OpenDRIVE: the concept of _rules_ came into place virtually only at standard 1.6, and several of the 1.6 rules can be applied also to previous version)_

Those two considerations make the overall topic quite tricky. We may reach a situation in which rules of newer standard may be applied to older standard, and we should provide enough flexibilty to allow to cover for this scenarios.

I think in this case we can steal the python concept of [`python_requires`](https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata-requires-python). The python requires is provided in the form of [version specifiers](https://packaging.python.org/en/latest/specifications/version-specifiers/#id5), which are useful to define a sequence of version for which something can be applied to. 

I suggest to use only the following comparison operator:

 * `<, >`: exclusive ordered comparison
 * `<=, >=`: inclusive ordered comparison

additionally, we will need a special **override** operator (i suggest `!`) to include at the end of the version specifier. It should be noted that specifers `==` and `!=` can be defined by using the proposed operators wisely.

We can effectively call the concept `APPLICABLE_VERSION` as suggested. The concept may work as follows (_for the sake of discussion, let's consider a rule uid for OpenDRIVE `asam.net:xodr:1.6.0:*`_):

 1. If no **applicable version** is provided, a rule is applied starting from the version declared in the uid, up to the most recent one 
    * _for the example, if the `APPLICABLE_VERSION=""`, then the rule should be applied to OpenDRIVEs of version 1.6.0, 1.6.1, 1.7.0, 1.8.0: the internal representation of the version specifier is something like `>=1.6.0`_
 
  2. If the **applicable version** is provided, and defines only an **upper bound**, then the version in the uid defines the lower bound 
     * _for the example, if the `APPLICABLE_VERSION="<1.8"`, then the rule should be applied to OpenDRIVEs of version 1.6.0, 1.6.1, 1.7.0: the internal representation of the version specifier is something like `>=1.6.0, <1.8.[max]`_

 3. If the **applicable version** is provided, and defines a lower bound, then the version in the uid defines the lower bound 
     * _for the example, if the `APPLICABLE_VERSION=">=1.5"`, then the rule should be applied to OpenDRIVEs of version 1.5.0, 1.6.0, 1.6.1, 1.7.0: the internal representation of the version specifier is something like `>=1.6.0, >=1.5.0`, and simplified in `>= 1.5.0`_
     
 4. The `!` operator discards the version in the uid, and covers the (I hope impossible) situations in which a rule defined on a specific standard, cannot be applied an that standard 
    * _for example: if the `APPLICABLE_VERSION="<=1.4,>=1.7!"`, then the rule should be applied to OpenDRIVEs of version 1.4.0, 1.7.0, 1.8.0: the internal representation of the version specifier is something like `<=1.4.[max],>=1.7.0`, and the `>= 1.6.0` component is completely discarded_

---

| Case Number | Example Rule            | `APPLICABLE_VERSION` | Internal Representation          | File versions to be checked       |
|-------------|-------------------------|----------------------|----------------------------------|-----------------------------------|
| 1           | `asam.net:xodr:1.6.0:*` | `""`                 | `">=1.6.0"`                      | 1.6.0, 1.6.1, 1.7.0, 1.8.0        |
| 2           | `asam.net:xodr:1.6.0:*` | `"<1.8"`             | `">=1.6.0,<1.8.[max]"`           | 1.6.0, 1.6.1, 1.7.0               |
| 3           | `asam.net:xodr:1.6.0:*` | `">=1.5"`            | `">=1.5.0,>=1.6.0" => ">=1.5.0"` | 1.5.0, 1.6.0, 1.6.1, 1.7.0, 1.8.0 |
| 4           | `asam.net:xodr:1.6.0:*` | `"<=1.4,>=1.7!"`     | `"<=1.4.[max],>=1.7.0"`          | 1.4.0, 1.7.0, 1.8.0               |


Possible implementation reference: [semver3](https://python-semver.readthedocs.io/en/latest/usage/compare-versions-through-expression.html)

Reply:
Thanks @MatteoRagni . I added the concept of applicable version range in this PR. It mostly follows your example, with few deviations to keep it simpler for the first implementation. The concept is explained in details in the documentation in the last commit.
1. `>, <, >=, <=` are supported.
2. Override `!` is not supported for now.
3. Comma separator, e.g., `APPLICABLE_VERSION = ">1.5.0,<=1.8.0"` acts as logical "AND", so that it's aligned with the Python version specifiers.
4. The concept is summarized as follows (which cover Case Number 1, 2 and 3 in your examples): If `APPLICABLE_VERSION` version is defined and contains at least one lower bound, then ignore definition setting in rule UID. Otherwise, definition setting in rule UID acts as lower bound.

Reply:
There is something that bothers me, which I completely missed in the previous comment: 

Do we want to include the version and "applicable version" concepts only in the OpenDRIVE checker? Is the comparison logic something useful enough to be considered for promotion in base-lib?

Reply:
Thanks @MatteoRagni for the review. I forgot to mention that the current concept is only implemented for OpenDrive. After the first release, the whole concept of "executor", in which "applicable versions" is part of, can be moved to baselib (with some revision).

We didn't want to introduce the "executor" concept earlier in the baselib as it was not stable, and it still contains few standard-specific dependencies for now. We can revise the concept before introducing it as it will govern how future checker bundles will be implemented.

Reply:
It's for me ok to postpone the generic concept to the next release and do it now only for OpenDRIVE, because we need to come to a point...

@hoangtungdinh : Can you please create a feature ticket for that? Not sure if it is more framework related, because it needs probably modifications in all our repository, maybe also documentation. But there is also the ""executor" concept", which is more implementation (baselib-py) releated. Feel free to create one ticket, which fits best to your thoughts at the moment. Thx.